From: Zander Brown <zbrown@gnome.org>
Date: Mon, 5 Jun 2023 14:37:25 +0100
Subject: thumbnail: support %m expansion

This provides the MIME the thumbnailer was invoked for to the
thumbnailer itself.

See: https://gitlab.freedesktop.org/xdg/xdg-specs/-/merge_requests/99

Origin: https://gitlab.gnome.org/GNOME/gnome-desktop/-/commit/320933c
---
 libgnome-desktop/gnome-desktop-thumbnail-script.c | 46 ++++++++++++++---------
 libgnome-desktop/gnome-desktop-thumbnail-script.h |  7 ++--
 libgnome-desktop/gnome-desktop-thumbnail.c        |  6 ++-
 3 files changed, 38 insertions(+), 21 deletions(-)

diff --git a/libgnome-desktop/gnome-desktop-thumbnail-script.c b/libgnome-desktop/gnome-desktop-thumbnail-script.c
index 068e913..6fca21c 100644
--- a/libgnome-desktop/gnome-desktop-thumbnail-script.c
+++ b/libgnome-desktop/gnome-desktop-thumbnail-script.c
@@ -69,11 +69,12 @@ typedef struct {
 
 static char *
 expand_thumbnailing_elem (const char *elem,
-			  const int   size,
-			  const char *infile,
-			  const char *outfile,
-			  gboolean   *got_input,
-			  gboolean   *got_output)
+                          const int   size,
+                          const char *mime_type,
+                          const char *infile,
+                          const char *outfile,
+                          gboolean   *got_input,
+                          gboolean   *got_output)
 {
   GString *str;
   const char *p, *last;
@@ -112,6 +113,10 @@ expand_thumbnailing_elem (const char *elem,
 	g_string_append_printf (str, "%d", size);
 	p++;
 	break;
+      case 'm':
+        g_string_append (str, mime_type);
+        p++;
+        break;
       case '%':
 	g_string_append_c (str, '%');
 	p++;
@@ -734,9 +739,10 @@ add_flatpak (GPtrArray   *array,
 
 static char **
 expand_thumbnailing_cmd (const char  *cmd,
-			 ScriptExec  *script,
-			 int          size,
-			 GError     **error)
+                         ScriptExec  *script,
+                         int          size,
+                         const char  *mime_type,
+                         GError     **error)
 {
   GPtrArray *array;
   g_auto(GStrv) cmd_elems = NULL;
@@ -797,11 +803,12 @@ expand_thumbnailing_cmd (const char  *cmd,
       char *expanded;
 
       expanded = expand_thumbnailing_elem (cmd_elems[i],
-					   size,
-					   script->s_infile ? script->s_infile : script->infile,
-					   script->s_outfile ? script->s_outfile : script->outfile,
-					   &got_in,
-					   &got_out);
+                                           size,
+                                           mime_type,
+                                           script->s_infile ? script->s_infile : script->infile,
+                                           script->s_outfile ? script->s_outfile : script->outfile,
+                                           &got_in,
+                                           &got_out);
 
       g_ptr_array_add (array, expanded);
     }
@@ -1127,9 +1134,10 @@ print_script_debug (GStrv expanded_script)
 
 GBytes *
 gnome_desktop_thumbnail_script_exec (const char  *cmd,
-				     int          size,
-				     const char  *uri,
-				     GError     **error)
+                                     int          size,
+                                     const char  *uri,
+                                     const char  *mime_type,
+                                     GError     **error)
 {
   g_autofree char *error_out = NULL;
   g_auto(GStrv) expanded_script = NULL;
@@ -1141,7 +1149,11 @@ gnome_desktop_thumbnail_script_exec (const char  *cmd,
   exec = script_exec_new (uri, error);
   if (!exec)
     goto out;
-  expanded_script = expand_thumbnailing_cmd (cmd, exec, size, error);
+  expanded_script = expand_thumbnailing_cmd (cmd,
+                                             exec,
+                                             size,
+                                             mime_type,
+                                             error);
   if (expanded_script == NULL)
     goto out;
 
diff --git a/libgnome-desktop/gnome-desktop-thumbnail-script.h b/libgnome-desktop/gnome-desktop-thumbnail-script.h
index cbd6bbf..9935b4d 100644
--- a/libgnome-desktop/gnome-desktop-thumbnail-script.h
+++ b/libgnome-desktop/gnome-desktop-thumbnail-script.h
@@ -31,8 +31,9 @@
 
 GBytes *
 gnome_desktop_thumbnail_script_exec (const char  *cmd,
-				     int          size,
-				     const char  *uri,
-				     GError     **error);
+                                     int          size,
+                                     const char  *uri,
+                                     const char  *mime_type,
+                                     GError     **error);
 
 #endif /* GNOME_DESKTOP_THUMBNAIL_SCRIPT_H */
diff --git a/libgnome-desktop/gnome-desktop-thumbnail.c b/libgnome-desktop/gnome-desktop-thumbnail.c
index 02234b1..d419f5f 100644
--- a/libgnome-desktop/gnome-desktop-thumbnail.c
+++ b/libgnome-desktop/gnome-desktop-thumbnail.c
@@ -1143,7 +1143,11 @@ gnome_desktop_thumbnail_factory_generate_thumbnail (GnomeDesktopThumbnailFactory
     {
       GBytes *data;
 
-      data = gnome_desktop_thumbnail_script_exec (script, size, uri, error);
+      data = gnome_desktop_thumbnail_script_exec (script,
+                                                  size,
+                                                  uri,
+                                                  mime_type,
+                                                  error);
       if (data)
         {
           pixbuf = pixbuf_new_from_bytes (data, error);
